### PR TITLE
Respect MSRV (MAYBE_BOUND_IN_WHERE)

### LINKED ISSUE #12370: multiple_bound_locations false postitive: msrv < 1.15
### Summary

Rust 1.15.0 is the first release that allows `?Sized` bound in a where-clause. https://blog.rust-lang.org/2017/02/02/Rust-1.15.html#other-improvements

On older versions, one would need to write _all_ bounds in the angle bracketed generic parameter list (not always reasonable) or put `?Sized` in the angle bracketed generic parameter list and rest of the bounds in the where-clause.

### Lint Name

multiple_bound_locations

### Reproducer

```toml
# Cargo.toml

[package]
name = "repro"
version = "0.0.0"
rust-version = "1.14"
```

```rust
// src/lib.rs

pub fn f<T: ?Sized>()
where
    T: Iterator<Item = Vec<u8>>,
{
}
```

```console
$ cargo clippy
warning: bound is defined in more than one place
 --> src/lib.rs:3:10
  |
3 | pub fn f<T: ?Sized>()
  |          ^
4 | where
5 |     T: Iterator<Item = Vec<u8>>,
  |     ^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#multiple_bound_locations
  = note: `#[warn(clippy::multiple_bound_locations)]` on by default
```

Clippy's preference is for one of the following, neither of which is better.

```rust
// unreadable for a verbose bound
pub fn f<T: ?Sized + Iterator<Item = Vec<u8>>>()

// incompatible with msrv
pub fn f<T>()
where
    T: ?Sized + Iterator<Item = Vec<u8>>
```

### Version

```text
rustc 1.78.0-nightly (ef324565d 2024-02-27)
binary: rustc
commit-hash: ef324565d071c6d7e2477a195648549e33d6a465
commit-date: 2024-02-27
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by sanxiyn**
Fix #12370.

changelog: [`multiple_bound_locations`]: respect MSRV for combining `?Sized` bound

### CODE REVIEW

**Review by y21 — COMMENTED** [review]
+1 for having this be part of the `trait_bounds` pass, fits nicely with the other lints, and also nice seeing the `cannot_combine_maybe_bound` function get some use

**Comment by y21** [comment]
Hmm, looking more closely into the lints covered by this pass... Doesn't the [`type_repetition_in_bounds`](https://rust-lang.github.io/rust-clippy/master/index.html#/type_repetition_in_bounds) lint basically cover everything that this lint catches, but more thorough (as in, will also catch regular types and not just type parameters, though it is in the nursery category)?
Not necessarily related to what this PR addresses though, so that's for another time ^^

File: clippy_lints/src/trait_bounds.rs [hunk 1]
```diff
@@ -237,6 +265,60 @@ impl TraitBounds {
...
+                && p.origin == PredicateOrigin::GenericParam
+                && let bounds = p
+                    .bounds
+                    .iter()
+                    .filter(|b| !self.cannot_combine_maybe_bound(cx, b))
+                    .collect::<Vec<_>>()
+                && !bounds.is_empty()
+                && let Some((_, ident)) = p.bounded_ty.as_generic_param()
+            {
+                generic_params_with_bounds.insert(ident.name.as_str().to_owned(), ident.span);
```
[the remark this answers is not in the document]
Comment (sanxiyn) [reply clippy_lints/src/trait_bounds.rs:282]: Yes, that's better. Now done.

File: clippy_lints/src/trait_bounds.rs [hunk 2]
```diff
@@ -237,6 +265,60 @@ impl TraitBounds {
...
+
+        for bound in gen.predicates {
+            if let WherePredicate::BoundPredicate(p) = bound
+                && p.origin == PredicateOrigin::GenericParam
+                && let bounds = p
+                    .bounds
+                    .iter()
+                    .filter(|b| !self.cannot_combine_maybe_bound(cx, b))
+                    .collect::<Vec<_>>()
+                && !bounds.is_empty()
```
[the remark this answers is not in the document]
Comment (sanxiyn) [reply clippy_lints/src/trait_bounds.rs:279]: Done.

File: clippy_lints/src/trait_bounds.rs [hunk 3]
```diff
@@ -237,6 +265,60 @@ impl TraitBounds {
...
+                && let ident = p.lifetime.ident
+                && let Some(bound_span) = generic_params_with_bounds.get(ident.name.as_str())
+            {
+                let where_span = ident.span;
+                span_lint(
+                    cx,
+                    MULTIPLE_BOUND_LOCATIONS,
+                    vec![*bound_span, where_span],
+                    "bound is defined in more than one place",
+                );
```
[the remark this answers is not in the document]
Comment (sanxiyn) [reply clippy_lints/src/trait_bounds.rs:317]: Sure.

**Comment by J-ZhengLi** [comment]
Hey @sanxiyn , this is a ping from triage. There hasn't been any activity for some time, you're always welcome to ask questions (there's a [zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy) if you're interested). It's also fine if you don't have the time right now. We just want to check in and make sure that you're not stuck.

**Comment by J-ZhengLi** [comment]
Hey @sanxiyn , this is a ping from triage. There hasn't been any activity for some time, just want to check in and make sure that you're working on this PR.

**Comment by xFrednet** [comment]
Hey this is triage, I'm closing this due to inactivity. If you want to continue this implementation, you're welcome to create a new PR. Thank you for the time, you already put into this!

Interested parties are welcome to pick this implementation up as well :)

@rustbot label +S-inactive-closed -S-waiting-on-author -S-waiting-on-review

**Pull request closed without merging**